File: agent_backend/routers/pinterest.py
# ─────────────────────────────────────────────────────────────────────────────
# Olive Mode — Pinterest Router
# File: routers/pinterest.py
# ─────────────────────────────────────────────────────────────────────────────

from fastapi import APIRouter, Query
from services.pinterest_client import pinterest_get

router = APIRouter(prefix="/pinterest", tags=["Pinterest"])


@router.get("/trends/top")
async def get_top_trends(
    region:               str  = Query(default="US"),
    trend_type:           str  = Query(default="monthly"),
    limit:                int  = Query(default=20, ge=1, le=50),
    include_predictions:  bool = Query(default=True),
    include_demographics: bool = Query(default=True),
    genders: str | None        = Query(default="female"),
    ages:    str | None        = Query(default=None),
):
    """Top trending keywords on Pinterest filtered to fashion & beauty."""
    params: dict = {
        "limit":            limit,
        "include_keywords": True,
    }
    if genders:              params["genders"]              = genders.split(",")
    if ages:                 params["ages"]                 = ages.split(",")
    if include_demographics: params["include_demographics"] = True

    data   = await pinterest_get(f"/trends/keywords/{region}/top/{trend_type}", params)
    trends = data.get("trends", [])

    return {
        "region":     region,
        "trend_type": trend_type,
        "filters": {
            "genders": genders.split(",") if genders else "all",
            "ages":    ages.split(",")    if ages    else "all",
        },
        "total":  len(trends),
        "trends": [
            {
                "keyword":       t.get("keyword"),
                "is_trending":   t.get("is_trending_now", False),
                "weekly_counts": t.get("weekly_counts", {}).get("adjusted_trend_counts", []),
                "prediction":    t.get("trend_prediction") if include_predictions else None,
                "demographics":  _extract_demographics(t) if include_demographics else None,
            }
            for t in trends
        ],
    }


@router.get("/trends/featured")
async def get_featured_topics(region: str = Query(default="US")):
    """Pinterest editorially curated featured trend topics."""
    data   = await pinterest_get("/trends/topics/featured", {"region": region})
    topics = data.get("items", [])
    return {
        "region": region,
        "total":  len(topics),
        "topics": [
            {
                "id":          t.get("id"),
                "name":        t.get("name"),
                "description": t.get("description"),
                "image_url":   t.get("image_url"),
                "trend_url":   t.get("trend_url"),
            }
            for t in topics
        ],
    }


@router.get("/trends/categories")
async def get_trending_categories(region: str = Query(default="US")):
    """Trending product categories on Pinterest."""
    data       = await pinterest_get("/trends/product_categories/trending", {"region": region})
    categories = data.get("items", [])
    return {
        "region": region,
        "total":  len(categories),
        "categories": [
            {
                "id":        c.get("id"),
                "name":      c.get("name"),
                "momentum":  c.get("trend_score"),
                "image_url": c.get("image_url"),
            }
            for c in categories
        ],
    }


@router.get("/trends/keyword/{keyword}")
async def get_keyword_trend(
    keyword:    str,
    region:     str = Query(default="US"),
    trend_type: str = Query(default="monthly"),
):
    """Trend data for a specific keyword."""
    data   = await pinterest_get(
        f"/trends/keywords/{region}/top/{trend_type}",
        {"limit": 50, "include_keywords": True}
    )
    trends        = data.get("trends", [])
    keyword_lower = keyword.lower()
    matches       = [t for t in trends if keyword_lower in t.get("keyword", "").lower()]

    if not matches:
        return {
            "keyword": keyword,
            "region":  region,
            "found":   False,
            "message": f"'{keyword}' not found in top {trend_type} trends for {region}.",
            "trends":  [],
        }

    return {
        "keyword": keyword,
        "region":  region,
        "found":   True,
        "total":   len(matches),
        "trends": [
            {
                "keyword":       t.get("keyword"),
                "is_trending":   t.get("is_trending_now", False),
                "weekly_counts": t.get("weekly_counts", {}).get("adjusted_trend_counts", []),
                "prediction":    t.get("trend_prediction"),
            }
            for t in matches
        ],
    }


# ─── Helper ───────────────────────────────────────────────────────────────────

def _extract_demographics(trend: dict) -> dict:
    demo = trend.get("demographics", {})
    if not demo:
        return {}
    return {
        "ages": {
            b.get("name"): f"{round(b.get('ratio', 0) * 100, 1)}%"
            for b in demo.get("ages", [])
        },
        "genders": {
            b.get("name"): f"{round(b.get('ratio', 0) * 100, 1)}%"
            for b in demo.get("genders", [])
        },
    }

File: src/lib/api/pinterest.ts
// ─────────────────────────────────────────────────────────────────────────────
// Olive Mode — Pinterest API Service
// Place at: src/lib/api/pinterest.ts
// ─────────────────────────────────────────────────────────────────────────────

import { type TrendResult } from "./chat";
import { extractKeyword } from "./trends";

const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";

// ─── Top trends ───────────────────────────────────────────────────────────────

export interface TopTrendsParams {
  region?:               string;
  trend_type?:           "weekly" | "monthly" | "yearly";
  limit?:                number;
  include_predictions?:  boolean;
  include_demographics?: boolean;
  genders?:              string;
  ages?:                 string;
}

export interface TopTrendsResponse {
  region:     string;
  trend_type: string;
  filters:    { genders: string | string[]; ages: string | string[] };
  total:      number;
  trends:     TrendResult[];
}

export async function fetchTopTrends(
  params: TopTrendsParams = {}
): Promise<TopTrendsResponse | null> {
  try {
    const query = new URLSearchParams({
      region:               params.region               ?? "US",
      trend_type:           params.trend_type           ?? "monthly",
      limit:                String(params.limit         ?? 20),
      include_predictions:  String(params.include_predictions  ?? true),
      include_demographics: String(params.include_demographics ?? true),
      ...(params.genders && { genders: params.genders }),
      ...(params.ages    && { ages:    params.ages }),
    });

    const response = await fetch(
      `${API_BASE}/pinterest/trends/top?${query.toString()}`
    );
    if (!response.ok) return null;
    return response.json() as Promise<TopTrendsResponse>;
  } catch {
    console.error("[pinterest] fetchTopTrends failed");
    return null;
  }
}

// ─── Keyword trend ────────────────────────────────────────────────────────────

export interface KeywordTrendResponse {
  keyword: string;
  region:  string;
  found:   boolean;
  total?:  number;
  trends:  TrendResult[];
}

export async function fetchKeywordTrend(
  keyword:    string,
  region      = "US",
  trend_type  = "monthly"
): Promise<TrendResult[]> {
  try {
    const cleanKeyword = extractKeyword(keyword);
    console.log(`[pinterest] keyword: "${keyword}" → cleaned: "${cleanKeyword}"`);

    const encoded  = encodeURIComponent(cleanKeyword);
    const response = await fetch(
      `${API_BASE}/pinterest/trends/keyword/${encoded}?trend_type=${trend_type}&region=${region}`
    );

    if (!response.ok) {
      console.warn(`[pinterest] request failed with status ${response.status} — skipping`);
      return [];
    }

    const data = await response.json() as KeywordTrendResponse;
    return data.found ? data.trends : [];
  } catch {
    console.error("[pinterest] fetchKeywordTrend failed");
    return [];
  }
}

// ─── Featured topics ──────────────────────────────────────────────────────────

export interface FeaturedTopic {
  id:          string;
  name:        string;
  description: string;
  image_url:   string;
  trend_url:   string;
}

export interface FeaturedTopicsResponse {
  region: string;
  total:  number;
  topics: FeaturedTopic[];
}

export async function fetchFeaturedTopics(
  region = "US"
): Promise<FeaturedTopicsResponse | null> {
  try {
    const response = await fetch(
      `${API_BASE}/pinterest/trends/featured?region=${region}`
    );
    if (!response.ok) return null;
    return response.json() as Promise<FeaturedTopicsResponse>;
  } catch {
    console.error("[pinterest] fetchFeaturedTopics failed");
    return null;
  }
}

// ─── Trending categories ──────────────────────────────────────────────────────

export interface TrendingCategory {
  id:        string;
  name:      string;
  momentum:  number;
  image_url: string;
}

export interface TrendingCategoriesResponse {
  region:     string;
  total:      number;
  categories: TrendingCategory[];
}

export async function fetchTrendingCategories(
  region = "US"
): Promise<TrendingCategoriesResponse | null> {
  try {
    const response = await fetch(
      `${API_BASE}/pinterest/trends/categories?region=${region}`
    );
    if (!response.ok) return null;
    return response.json() as Promise<TrendingCategoriesResponse>;
  } catch {
    console.error("[pinterest] fetchTrendingCategories failed");
    return null;
  }
}